### PR TITLE
Fix Axelar wallets stuck syncing on pruned blocks

### PR DESCRIPTION
**Pull request by peachbits**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

none

### Description

Asana task: [Follow up on "Update Axelar and Osmosis rpc"](https://app.asana.com/1/9976422036640/task/1218494629884470)

Existing Axelar wallets with transaction history stopped syncing partway. QA saw two wallets pinned at 75% and 87.5%, a third at 25%, and a resync did not clear it.

A pruned RPC node causes it, not a cleaner. [EdgeApp/edge-currency-accountbased#1102](https://github.com/EdgeApp/edge-currency-accountbased/pull/1102) moved Axelar to `axelar-rpc.publicnode.com` (main) and `axelar-rpc.polkachu.com` (archive). Both prune: publicnode stores blocks from 2026-09-06 (about height 34.25M), polkachu from 2026-06-15. Publicnode's transaction index still returns these wallets' older transactions (heights 28,035,527 to 29,739,468), but it no longer holds those blocks, so `getBlock` answers `height 28842747 is not available, lowest height is 34246686`. That error left `queryTransactionsInner`, which ended `queryTransactions` before it recorded `archivedTxLastCheckTime`, so every 20 second loop and every resync stopped at the same step.

The displayed ratio is `(balance + history) / 2`, and a fresh or resynced wallet runs four history passes (two nodes, two queries) worth 0.25 each. That maps the reported numbers exactly:

| Wallet | Pass that threw | History | Displayed |
| --- | --- | --- | --- |
| `axelar1z4spcxyucu63n3l7u9r4qnwl5gw30cspmt7r5g` | publicnode `coin_spent.spender` | 0.5 | 75% |
| `axelar1z4ylygfj4t69g649g88lj6grs758z3qm57zgf6` | publicnode `coin_received.receiver` | 0.75 | 87.5% |

The cleaners were the first suspect, and reproduction ruled them out: `reduceCoinEventsForAddress` parsed every returned transaction correctly (and its caller already catches it), while cosmjs's Comet38 `assertNotEmpty(attribute.key)` never fired, since no returned event carried an empty key or a null value.

#### Changes

1. **A transaction already recorded at a block height is left alone**, since there is nothing to date or add. That also ends a standing waste: paging always restarts at page 1, so a query holding 120 rows made 120 `getBlock` calls on every pass, forever.
2. **A block the querying node has pruned is looked up on the archive nodes**, whose clients are created on demand. The sync lists archive nodes only every two weeks, while a transaction older than the main node's roughly 9 days of blocks can turn up on any pass, so on-demand setup is what keeps those rows datable. A transaction is skipped only when no node has its block, and any other failure still ends the pass, leaving it to run again later.
3. **A page the node keeps reporting as out of range is retried at most three times.** The retry called `continue` inside `do/while (true)`, which skipped the page increment, so such a node was asked for the same page again forever with no delay.
4. **Axelar's archive node points back at `axelar-archrpc.chainode.tech`.** [EdgeApp/edge-currency-accountbased#1102](https://github.com/EdgeApp/edge-currency-accountbased/pull/1102) replaced it after a connection failure from the device, but it answers again: earliest block 1 (2021-12-22), caught up, serving every block and `tx_search` result these wallets need. Polkachu cannot date anything older than June 2026.

The engine changes cover every Cosmos chain, so the next node that prunes or dies cannot pin a wallet below 100% this way.

#### Testing

- `verify-repo.sh . --base origin/master` passed, full suite included. Its `prepare` step runs this repo's `types` script, so the tree type-checks inside verification as well.
- Ten cases in `test/cosmos/queryTransactionsInner.test.ts` drive `queryTransactionsInner` through a fake engine: dating from an archive node when the querying node has pruned the block, skipping when no node can date it, ending the pass on any other block error, ending the pass when the archive nodes cannot be set up, looking up no block for an already recorded transaction, dating one held only as pending, keeping the stored watermark through a pass that adds nothing, offering no archive clients when a chain configures none, giving up on a page that stays out of range, and continuing when a retried page answers.
- Live run against the real nodes, driving `queryTransactionsInner` with only the main rpcNode: both wallets above complete with every row dated (2026-05-05, 2026-05-21, 2026-06-08 and 2026-09-15). The dates for the pruned heights come from the archive node through the on-demand lookup, which is the path this branch adds.

Left out of this branch: QA's third wallet at 25% cannot be mapped without its address, because a wallet with no enabled tokens and a working balance query can only show 50%, 75% or 87.5% from this failure.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes core Cosmos transaction-history sync and RPC fallback behavior across all Cosmos chains; incorrect watermark or archive logic could miss or duplicate history, but behavior is heavily unit-tested.
> 
> **Overview**
> Fixes **Cosmos wallets (including Axelar) stuck partway through sync** when an RPC node still returns old transactions from `tx_search` but no longer serves the corresponding blocks.
> 
> **CosmosEngine** now resolves block times through **`queryBlockDate`**, falling back to **on-demand archive RPC clients** when the querying node reports a pruned height, and **skips** a transaction only when no configured node can date it—while **holding the sync watermark** so undated rows are retried on a later pass. It also **avoids re-dating transactions already confirmed** in the wallet (`hasConfirmedTx`, with txid normalization), and **caps retries** when a node keeps reporting the same `tx_search` page as out of range (`MAX_PAGE_RETRIES`).
> 
> **Axelar** archive configuration switches back to **`axelar-archrpc.chainode.tech`** so pre–June 2026 history can be dated. New unit tests in **`queryTransactionsInner.test.ts`** cover archive dating, watermark behavior, paging retries, and related edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 72c612482953f8133ce87a413e34c3a372d00305. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/cosmos/engine/CosmosEngine.ts [hunk 1]
```diff
@@ -843,13 +871,81 @@ export class CosmosEngine extends CurrencyEngine<
...
+      } catch (error: unknown) {
+        continue // processCosmosTransaction ignores unknown denoms as well
+      }
+      const index = this.findTransaction(tokenId, txid)
+      if (index < 0) return false
+      const tx = this.transactionList[tokenId ?? '']?.[index]
+      if (tx == null || tx.blockHeight <= 0) return false
+    }
+    return true
+  }
```
[the remark this answers is not in the document]
Comment (peachbits) [reply src/cosmos/engine/CosmosEngine.ts:962]: Valid, fixed in d9287d74. hasConfirmedTx now normalizes before the lookup, matching how the map is keyed: CurrencyEngine stores txIdMap[safeTokenId][normalizeAddress(tx.txid)], and MoneroEngine normalizes at its call site the same way. Checked directly against the engine: with a lowercase key stored, the uppercase txid now resolves, and an unknown txid still misses.

The test fake was why this slipped through. It indexed txids exactly as passed, so an uppercase lookup matched there while the real map never would. It now keys on the lowercase form, so the old behavior fails the test.

On the second half of the report: the backward move of newestTxid is limited to rows not yet recorded. With the skip working, such rows are skipped cheaply on the following pass instead of being re-dated, so it converges rather than costing repeated lookups.

**Review by j0ntz — APPROVED** [review]

**Review by j0ntz — APPROVED** [review]

**Comment by Metavcl** [comment]
Nice fix. The deeper failure is retention asymmetry: `tx_search` returns a transaction at height H, but `block(H)` is already pruned. An “archive” label is not sufficient unless these two data paths are consistent.

For wallet and accounting workloads, an RPC provider should publish machine-readable retention metadata such as:

* earliest block available
* earliest transaction-index height
* pruning schedule
* archive/full-node mode

An archive acceptance test should also sample transactions near the retention boundary and verify that every indexed transaction can resolve its block and timestamp. If the index intentionally outlives block storage, the two retention floors should be documented separately.

The client behavior here is also correct: reuse already-recorded transactions, fall back only on a typed pruned-block error, bound repeated page failures, and report incomplete history instead of treating missing data as zero or retrying the same page every 20 seconds.

BlockPI’s [[Dedicated Node](https://blockpi.io/dedicated-node/)](https://blockpi.io/dedicated-node/) service supports configurable Full/Archive deployments across 70+ networks. For wallet infrastructure, retention depth and index-to-block consistency can be defined as deployment acceptance requirements.

For RPC package purchases, use discount code `buildwithblockpi`.
